src/openstack_cloud/openstack_cloud.py

```diff
@@ -0,0 +1,608 @@
+from openstack.compute.v2.server import Server as OpenstackServer
+from openstack.connection import Connection as OpenstackConnection
+from openstack.network.v2.security_group import SecurityGroup as OpenstackSecurityGroup
+from paramiko.ssh_exception import NoValidConnectionsError
+
+from errors import KeyfileError, OpenStackError, SSHError
+
+logger = logging.getLogger(__name__)
+
+_CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
+
+# Update the version when the security group rules are not backward compatible.
+_SECURITY_GROUP_NAME = "github-runner-v1"
+
+_CREATE_SERVER_TIMEOUT = 5 * 60
+_SSH_TIMEOUT = 30
+_SSH_KEY_PATH = Path("/home/ubuntu/.ssh")
+_TEST_STRING = "test_string"
+
+
+@dataclass
+class OpenstackInstance:
+    """Represents an OpenStack instance.
+
+    Attributes:
+        server_id: ID of server assigned by OpenStack.
+        server_name: Name of the server on OpenStack.
+        instance_id: ID used by OpenstackCloud class to manage the instances. See docs on the
+            OpenstackCloud.
+        addresses: IP addresses assigned to the server.
+        status: Status of the server.
+    """
+
+    server_id: str
+    server_name: str
+    instance_id: str
+    addresses: list[str]
+    status: str
+
+    def __init__(self, server: OpenstackServer, prefix: str):
+        """Construct the object.
+
+        Args:
+            server: The OpenStack server.
+            prefix: The name prefix for the servers.
+
+        Raises:
+            ValueError: Provided server should not be managed under this prefix.
+        """
+        self.server_id = server.id
+        self.server_name = server.name
+        self.status = server.status
+        self.addresses = [
+            address["addr"]
+            for network_addresses in server.addresses.values()
+            for address in network_addresses
+        ]
+
+        if not self.server_name.startswith(f"{prefix}-"):
+            # Should never happen.
+            raise ValueError(
+                f"Found openstack server {server.name} managed under prefix {prefix}, contact devs"
+            )
+        # Disable E203 (space before :) as it conflicts with the formatter (black).
+        self.instance_id = self.server_name[len(prefix) + 1 :]  # noqa: E203
+
+
+@contextmanager
+def _get_openstack_connection(
+    clouds_config: dict[str, dict], cloud: str
+) -> Iterator[OpenstackConnection]:
+    """Create a connection context managed object, to be used within with statements.
+
+    The file of _CLOUDS_YAML_PATH should only be modified by this function.
+
+    Args:
+        clouds_config: The configuration in clouds.yaml format to apply.
+        cloud: The name of cloud to use in the clouds.yaml.
+
+    Raises:
+        OpenStackError: if the credentials provided is not authorized.
+
+    Yields:
+        An openstack.connection.Connection object.
+    """
+    if not _CLOUDS_YAML_PATH.exists():
```

Comment:
Check what we do if the cloud configuration changes

Reply:
Update the code and added comment.